### PR TITLE
Issues/broker dev/keychain store tomb stone

### PR DESCRIPTION
**Pull request by jasoncoolmax**
Referring #445 
Tombstone feature added to ADKeychainTokenCacheStore.Now if an item will be marked as dead and left in keychain rather than being deleted.

In particular, the following changes are made:

1) property *correlationId* is added to ADTokenCacheStoreItem. If the token is tombstoned, the value will be the correlation ID of the request that we got the error from; if not, it is the correlation ID of the request through which we get the token.

2) property *markAsDead* is added to ADTokenCacheStoreItem. It shows whether the token is tombstoned.

3) unit tests for tombstone feature is added.

### CODE REVIEW

File: ADALiOS/ADALiOS/ADTokenCacheStoring.h [hunk 1]
```diff
@@ -66,4 +66,27 @@
                    userId:(NSString *)userId
                     error:(ADAuthenticationError * __autoreleasing *)error;
 
+/*! Mark a cache item as dead instead of removing it.
+ @param item: the item being marked as dead. Key will be extracted from the item.
+ @param userId: The user for which the item will be removed. Can be nil, in which case items for all users with
+ the specified key will be removed.
+ @param error: in case of an error, if this parameter is not nil, it will be filled with
+ the error details. */
+- (void)tombstoneItem:(ADTokenCacheStoreItem *)item
```
Comment (RPangrle) [inline ADALiOS/ADALiOS/ADTokenCacheStoring.h]: This is a public interface, so anything you add to it also becomes public. Tombstoning is more a private implementation detail that we should minimize the exposure of.

File: ADALiOS/ADALiOS/ADTokenCacheStoreItem.h [hunk 2]
```diff
@@ -46,6 +46,13 @@
 
 @property (retain) ADUserInformation* userInformation;
 
+/*! Correlation ID of the request through which we get the token.
+ If the token is tombstoned, the value will be the correlation ID of the request that we got the error from.  */
+@property (retain) NSString* correlationId;
+
+/* Whether the token is tombstoned. */
+@property BOOL markAsDead;
```
Comment (RPangrle) [inline ADALiOS/ADALiOS/ADTokenCacheStoreItem.h]: This is a verb phrase, a property should be a noun, preferably just "tombstone."

File: ADALiOS/ADALiOS/ADAuthenticationContext+TokenCaching.m [hunk 3]
```diff
@@ -94,8 +95,8 @@ - (ADTokenCacheStoreItem*)findCacheItemWithKey:(ADTokenCacheStoreKey*) key
         }
         else
         {
-            //We have a cache item that cannot be used anymore, remove it from the cache:
-            [self.tokenCacheStore removeItemWithKey:key userId:userId.userId error:nil];
+            //We have a cache item that cannot be used anymore, mark it as dead in the cache:
```
Comment (RPangrle) [inline ADALiOS/ADALiOS/ADAuthenticationContext+TokenCaching.m]: I wouldn't bother tombstoning expired ATs. There's no question why those are gone.

File: ADALiOS/ADALiOS/ADKeychainTokenCacheStore.m [hunk 4]
```diff
@@ -258,12 +269,14 @@ - (ADTokenCacheStoreItem*)itemFromKeyhainAttributes:(NSDictionary*)attrs
     for (NSDictionary* attrs in items)
     {
         ADTokenCacheStoreItem* item = [self itemFromKeyhainAttributes:attrs];
-        if (!item)
+        if (item)
         {
-            continue;
+            BOOL validity =  fromGraveyard ? [item markAsDead] : ![item markAsDead];
```
Comment (RPangrle) [inline ADALiOS/ADALiOS/ADKeychainTokenCacheStore.m]: Let higher layers of code decide whether or not they want to deal with tombstoned items.

File: ADALiOS/ADALiOS/ADTokenCacheStoring.h (hunk 1 above)
Comment (jasoncoolmax) [inline ADALiOS/ADALiOS/ADTokenCacheStoring.h]: Oh, I see, it is a private implementation. I will change it :)

File: ADALiOS/ADALiOS/ADTokenCacheStoreItem.h (hunk 2 above)
Comment (jasoncoolmax) [inline ADALiOS/ADALiOS/ADTokenCacheStoreItem.h]: Sure :)

**Comment by RPangrle** [comment]
Also, make sure you replace the refresh token with something that makes it obvious it's a tombstone, ie. "<tombstone>" followed by a short description of why it's tombstoned (the server error we got when it was rejected, or manually removed by <insert bundle id here>) This way if there are ADAL apps around that haven't been updated even if they try to use it the auth request will still fail (albeit not as gracefully as if they updated)

**Comment by jasoncoolmax** [comment]
This branch for tombstone implementation is too old to be merged.

**Pull request closed without merging**